---
title: 'BioKanga – a general purpose bioinformatics toolkit and reproducible evaluation frameworks for RNA- and DNA-Seq aligners'
author:
- Author First (University of Foo)
- Another Person (University of Bar)
output:
  bookdown::pdf_document2:
    number_sections: yes
    toc: false
abstract:
  This is the abstract.
bibliography: references.bib
csl: elsevier-harvard.csl
link-citations: true
urlcolor: blue
linkcolor: red
---

```{r, echo = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  fig.path = "figures/"
)
```


# Summary

[//]: # (Cite papers like this: `[@Kim2015]` [@Kim2015; @Baruzzo2016] matching entries in `references.bib`.)

BioKanga is a general purpose bioinformatics toolkit targeting many utilised within bioinformatics workflows.
We present BioKanga toolkit in the context of a turn-key reproducible and easily extendible evaluation frameworks for RNA-Seq and DNA-Seq.
We demonstrate how BioKanga align, a general purpose short read aligner performs on par with specialist RNA-Seq aligners.
We also show how design decisions behind BioKanga align lead to more robust results then those produced by popular alignment tools such as bwa or bowtie2 when aligning DNA-Seq reads to large, complex and repetitive polyploid genomes.

## BioKanga toolkit

Individual BioKanga tools are integrated within a single executable process.
The tools are presented as user selected submodules, with each submodule exposing a set of parameterisation defaults which can be overridden as appropriate given a priori knowledge.
Although a general purpose toolkit, the majority of individual BioKanga submodules have proven to be very competitive with other specialised  programs individually targeting individual bioinformatics tasks. BioKanga is developed in C and is available at [https://github.com/csiro-crop-informatics/biokanga](https://github.com/csiro-crop-informatics/biokanga).
This paper gives an overview of the `align` submodule and evidences the competiveness of the BioKanga align submodule relative to one of the specialised RNA-seq aligners, HISAT2 [@Kim2015], as recently benchmarked by Giacomo Baruzzo et al. [@Baruzzo2016].

## BioKanga align within RNA-Seq benchmarking framework

A major task in most NGS dataset dependent bioinformatics workflows is the alignments of the reads against targeted sequences.
Such targets may be in the form of a reference fully assembled genome comprising multiple chromosomal sized sequences, partially assembled genomes comprising scaffolded contigs, transcribed gene sequences, or in fact any arbitrary set of sequences having some experimental relevance.
The NGS datasets may themselves have been generated using a variety of technological resources with different library preparation protocols for which there will be different optimal best alignment strategies required.
The biokanga alignment submodule, `biokanga align`,exposes a large number of internal thresholds and alignment strategies allowing the experimenter to select the most optimal set of parameters which maximise their chances of meeting or exceeding experimental objectives.
It must be noted that there are defaults for most of the parameterisation thresholds which have been chosen empirically and in most cases can deliver alignments which optimally target experimental objectives.
In the benchmarking experiment published by Giacomo Baruzzo et al. [@Baruzzo2016], two sets (human and malaria) of 100bp paired end BEERS [@Grant2011] simulated reads containing 3 different complexities of error rates were generated with these sets consisting of simulated reads drawn from merged combinations of selected known transcriptomes wherein exons of individual transcripts have been mapped on to the genomic DNA assemblies.
In addition to the simulated reads the BEERS [@Grant2011] reads simulator also outputs the genomic DNA loci of where individual reads were sourced from including splice junctions spanned by these reads with this information defined as comprising the ‘ground truth’.
Alignments were generated by each of the benchmarked aligners using the total genomic DNA assemblies as targets with the loci of resultant alignments then compared against the BEERS reported ground truth loci.
From this comparison, various statistics by the Giacomo Baruzzo et al [@Baruzzo2016] benchmarking framework are then reported including proportions (incorrectly/correctly) of reads aligned, bases aligned, splice junctions spanned and `microInDels` identified.
Using the same benchmarking framework [@Baruzzo2016] and BEERS simulated reads, BioKanga align was used to compare the qualitative performance of using the general purpose BioKanga toolkit aligner, using RNA-seq parameterisation, against the specialised RNA-seq aligner HISAT2 [@Kim2015] used in the original benchmarking [@Baruzzo2016].

The BEERS simulated reads have a number of deficiencies in that these generated reads do not realistically approximate reads with the error profiles as normally observed in Illumina PE RNA-seq sequenced datasets but decided to use the BEERS simulated reads so that the unchanged benchmarking framework result sets as published by Baruzzo et al. {ref}, can be used for comparison with Biokanga alignments.
One change which was made was to the supplementary simulated read sets incorporating partially retained adapters.
These retained adapters were only incorporated at the 3' end of reads (retention is normally at the 5'), were of fixed length (multiples of 10), only the 1st 1 million (10%) read pairs had retained adapters and both ends of these read pairs had exactly the same length retained adapter.
The retained adapter reads were generated by the `add_adapter2fasta_V2.pl` Perl script in the benchmarking framework.
A new script was written with a more realistic distribution of retained adaptor lengths with adaptors, where retained, only at the 5' end of reads.
The hisat2 (as an exemplary specialised RNA-seq aligner used in the original benchmarking) plus BioKanga were then benchmarked against these improved retained adapter datasets.

# Retained Adaptors and Artefact Sequences Confound Aligners

It is very common for NGS reads to contain artefact sub-sequences at the 5' and/or 3' read ends which have originated as a by-product of library preparation prior to actual base calling or inadequate adaptor trimming post sequencing by the sequencing dataset provider.
A combination of methods are generally employed for reduction of these artefact sub-sequences prior to alignment of reads against the targeted genome assembly. With RNA-seq reads the target may be either genomic or transcriptomic.
The generally employed methods use a set of known artefact sequences, e.g. adaptor sequences, which when detected at the 5' or 3' of a read will result in that read being hard trimmed until the known artefact sequence is no longer detectable.
Some methods simply hard trim a fixed number of nucleotides off. The question naturally arises as to what K-mer length of a putative artefact sequence must be present at the 5' or 3' end of a read before there can be confidence that the K-mer sequence of nucleotides is part of an artefact sequence rather than a naturally occurring K-mer sequence within the targeted assembly or transcriptome.
Additionally, sequenced artefacts will have the same background base call error rate profile as the sequenced targeted fragments.
Thus, it is common to require a relatively large K-mer of known artefact sequence to exactly match at the 5' or 3' end of a read before trimming that read to decrease the false positive rate.
Post trimming, during alignments to the targeted genome or transcriptome, a high substitution rate is required to gain alignments with those reads still containing end artefacts less than the K-mer detection threshold and/or which contained base calling substitution errors.
A high substitution rate may also result in reads which did not contain end artefact sequences being misaligned unless care is taken to only select alignments with minimal required substitutions.
A further confounding factor is that the artefact nucleotide sequences themselves may not be known as would be the case with cross over or chimeric joining during library PCR.

# BioKanga Utilises Adaptive Chimeric Trimming
BioKanga is unique amongst contemporary aligners in that it provides the implementation of a method whereby alignments are made without the need to firstly detect and trim end artefact sequences before actual alignment processing.
The degree of end trimming is adaptive so as to maximise the read alignment length whilst minimising the number of required substitutions along the alignment length.
This combination of maximising alignment length together with minimising required substitutions maximises confidence in the resultant accepted alignments.
here is no requirement to have prior knowledge of where in the sequencing phases the artefacts were introduced or the actual nucleotide sequences of these artefacts.
 The only constraint is that the artefacts must be at the 5' and/or the 3' ends of reads.
BioKanga will explore all putative alignments for a given read and select as aligned that alignment, from all putative, having maximal alignment length with minimal substitutions over the alignment length.
Maximal alignment length is prioritised over minimal substitutions.
With paired end read datasets the reads individually can be adaptively chimeric trimmed before normal paired end processing commences.

The relative success of this approach is exemplified by the performance of BioKanga within a benchmark framework originally targeting RNA-seq aligners whereby BioKanga accurately aligns more reads than HiSat2 which was one of the top performing aligners in the original study.
Adaptive end trimming enables BioKanga to trim RNA-seq reads back over splice junctions into exons when aligning RNA-seq against genomic targets containing introns.

# Adaptive Chimeric Trimming Method

BioKanga will first attempt to align reads without any trimming using the maximum number of allowed substitutions as specified by the user.
If there are multiple full length putative alignments for a given read then BioKanga will accept only that alignment requiring the least number of substitutions.

Reads remaining unaligned are then processed as follows:

a) $K = L/(S-1)$, where L is read length and $S$ is the number of allowed substitutions proportional to the read length, specifies the length of core sub-sequences to be segmented from each read. A given read will provide $N = (L+K-1)/K$ K-mer cores, with the Nth 3' right hand core partially overlapping with the previous core when $(L+K-1)/K$ is not integral.
b) Cores are iterated from 1 (5' left most) up to N (3' right most) with each K-mer core exact matched, no substitutions allowed, against the targeted sequences. A given read may have zero or more core K-mers with exact matches, and each matching core K-mers may be aligning to multiple target loci.
c) Each exactly matching K-mer core is then extended left and right within the boundaries of the originating read length with extension stopped after accumulating mismatches up to the maximum number of allowed substitutions. All alignment loci, sense, alignment length, and number of substitutions required are noted for each K-mer core.
d) The original unaligned read is then reverse complemented to enable antisense alignments to targeted sequences and steps b) through c) inclusive are repeated to derive both sense and antisense core extensions.
e) The best alignment loci chosen for a given read will be that extended K-mer core in each read which resulted in an alignment with maximal extended length and having the least required substitutions proportional to the alignment extended length.

The net result of the forgoing processing is that reads with end artefact sub-sequences will be adaptively trimmed without requiring prior knowledge of these artefacts or their nucleotide sequences whilst minimising the number of nucleotides required to be trimmed from each read.

# Benchmark Methods

Benchmarking procedure of Baruzzo et al. is very thorough and extensive, but the implementation is highly rigid with many hard-coded paths and the overall design making it difficult to port, adapt and even re run.
To make our results more readily reproducible, we have developed a Nextflow [@DiTommaso2017] pipeline which seamlessly handles tasks ranging from data download through execution of the aligners to summarising the results.
A key benchmarking step relying on the original procedure has been wrapped in a Docker container and is executed by the Nextflow pipeline using Singularity software.
The Nextflow pipeline automatically downloads test datasets (complexity levels T1, T2, T3, human only) generated with the BEERS simulator and made available by Giacomo Baruzzo et al. at http://bioinf.itmat.upenn.edu/BEERS/bp1/datasets.php.
Complexity level T1 is documented as having a substitution rate of 0.001, indel rate of 0.0005 and error rate 0.005.
Complexity level T2 is documented as having increased substitution rates of 0.005, indel rates of 0.002 and error rates 0.01.
Complexity level T3 further increased substitution rates to 0.03, indel rates to 0.005 and error rates to 0.02.
T3 also has an error rate of 0.5 in the last (3') 0 bases.
It is also stated that approximately 40% of reads involve introns with PE insert size distributions having minimum length 100bp, mean 200bp and maximum of 500 bp.
The benchmarking framework was downloaded from [http://bioinf.itmat.upenn.edu/BEERS/bp1/scripts/Alignment_scripts.zip](http://bioinf.itmat.upenn.edu/BEERS/bp1/scripts/Alignment_scripts.zip) and [http://bioinf.itmat.upenn.edu/BEERS/bp1/scripts/aligner_benchmark_khayer.zip](http://bioinf.itmat.upenn.edu/BEERS/bp1/scripts/aligner_benchmark_khayer.zip) in accordance with the download instructions presented by the benchmarking authors  at [http://bioinf.itmat.upenn.edu/BEERS/bp1/software.php](http://bioinf.itmat.upenn.edu/BEERS/bp1/software.php)

[//]: # (HISAT2 2.0.5 released 04/11/2016 binary was downloaded from  [ftp://ftp.ccb.jhu.edu/pub/infphilo/hisat2/downloads/hisat2-2.0.5-Linux_x86_64.zip]([ftp://ftp.ccb.jhu.edu/pub/infphilo/hisat2/downloads/hisat2-2.0.5-Linux_x86_64.zip))

[//]: # (The BioKanga 4.3.3 integrated toolkit was added as a new job in the benchmarking framework `./jobs` subdirectory and as a new template to the framework `./templates subdirectory`.)

The `master.rb` Ruby script was modified to include biokanga as another aligner to be benchmarked.
A new script `add_adapter2fasta_V3.pl` was developed to fix the issues with adaptor retentions experienced when using datasets generated by the pre-existing `add_adapter2fasta_V2.pl`.



# Acknowledgements {-}

[https://github.com/Robinlovelace/rmarkdown-paper-repo](https://github.com/Robinlovelace/rmarkdown-paper-repo)

# References

